plugins/meta/portmap/portmap.go

```diff
@@ -0,0 +1,266 @@
+		rule := []string{
+			"-p", entry.Protocol,
+			"--dport", strconv.Itoa(entry.HostPort)}
+
+		if entry.HostIP != "" {
+			rule = append(rule,
+				"-d", entry.HostIP)
+		}
+
+		rule = append(rule,
+			"-j", "DNAT",
+			"--to-destination", fmtIpPort(containerIP, entry.ContainerPort))
+
+		out = append(out, rule)
+	}
+	return out
+}
+
+// genToplevelSnatChain creates the top-level summary snat chain.
+// IMPORTANT: do not change this, or else upgrading plugins will require
+// manual intervention
+func genToplevelSnatChain(isV6 bool) chain {
+	return chain{
+		table: "nat",
+		name:  TopLevelSNATChainName,
+		entryRule: []string{
+			"-s", localhostIP(isV6),
+			"!", "-d", localhostIP(isV6),
+		},
+		entryChains: []string{"POSTROUTING"},
+	}
+}
+
+// genSnatChain creates the snat (localhost) chain for this container.
+func genSnatChain(netName, containerID string) chain {
+	name := formatChainName("SN-", netName, containerID)
+	comment := fmt.Sprintf(`snat name: "%s" id: "%s"`, netName, containerID)
+
+	return chain{
+		table: "nat",
+		name:  name,
+		entryRule: []string{
+			"-m", "comment",
+			"--comment", comment,
+		},
+		entryChains: []string{TopLevelSNATChainName},
+	}
+}
+
+// snatRules sets up masquerading for connections to localhost:hostport,
+// rewriting the source so that returning packets are correct.
+func snatRules(entries []PortMapEntry, containerIP net.IP) [][]string {
+	isV6 := (containerIP.To4() == nil)
+
+	out := make([][]string, 0, len(entries))
+	for _, entry := range entries {
+		out = append(out, []string{
+			"-p", entry.Protocol,
+			"-s", localhostIP(isV6),
+			"-d", containerIP.String(),
+			"--dport", strconv.Itoa(entry.ContainerPort),
+			"-j", "MASQUERADE",
+		})
+	}
+	return out
+}
+
+// enableLocalnetRouting tells the kernel not to treat 127/8 as a martian,
+// so that connections with a source ip of 127/8 can cross a routing boundary.
+func enableLocalnetRouting(ifName string) error {
+	routeLocalnetPath := "net.ipv4.conf." + ifName + ".route_localnet"
+	_, err := sysctl.Sysctl(routeLocalnetPath, "1")
+	return err
+}
+
+// unforwardPorts deletes any iptables rules created by this plugin.
+// It should be idempotent - it will not error if the chain does not exist.
+func unforwardPorts(config *PortMapConf) error {
+	// Teardown v4
+	ipt, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	if err != nil {
+		return err
+	}
+	dnatChain := genDnatChain(config.Name, config.ContainerID, nil)
+	if err := dnatChain.teardown(ipt); err != nil {
+		return fmt.Errorf("could not teardown ipv4 dnat: %v", err)
+	}
+	snatChain := genSnatChain(config.Name, config.ContainerID)
+	if err := snatChain.teardown(ipt); err != nil {
+		return fmt.Errorf("could not teardown ipv4 snat: %v", err)
+	}
+
+	ip6t, err := iptables.NewWithProtocol(iptables.ProtocolIPv6)
+	if err != nil {
+		return err
+	}
+
+	if err := dnatChain.teardown(ip6t); err != nil {
+		return fmt.Errorf("could not teardown ipv6 dnat: %v", err)
```

Comment:
Hmm, I think we want a way to disable ip6tables calls.  There are still a number of systems which don't come with ip6 support today (e.g. GCI)

Specifically we're seeing this when we try out the plugin on GCI:

```
NetworkPlugin cni failed to teardown pod \"fluentd-gcp-v2.0-h101m_kube-system\" network: could not teardown ipv6 dnat: exit status 3: ip6tables v1.4.21: can't initialize ip6tables table `nat':
```

Some rather hastily thought out options:
- Emit error messages but continue DEL processing as normal here
- More fine-grained error handling to determine if cleanup failed, or if it's just not supported.
- A switch which can be set to enable/disable ipv6 in the plugin entirely.